### PR TITLE
Record WAL metrics

### PR DESCRIPTION
**Pull request by xiang90**

### CODE REVIEW

**Comment by xiang90** [comment]
@yichengq update

File: wal/wal.go [hunk 1]
```diff
@@ -317,7 +327,11 @@ func (w *WAL) sync() error {
 			return err
 		}
 	}
-	return w.f.Sync()
+	start := time.Now()
+	err := w.f.Sync()
+	syncHdr.RecordValue(time.Since(start).Nanoseconds() / int64(time.Microsecond))
+	metrics.Counter("wal.syncs").Add()
```
Comment (yichengq) [inline wal/wal.go]: syncOps

File: wal/wal.go [hunk 2]
```diff
@@ -269,6 +278,7 @@ func (w *WAL) ReadAll() (metadata []byte, state raftpb.HardState, ents []raftpb.
 	// create encoder (chain crc with the decoder), enable appending
 	w.encoder = newEncoder(w.f, w.decoder.lastCRC())
 	w.decoder = nil
+	metrics.Gauge("wal.index").Set(int64(w.enti))
```
Comment (yichengq) [inline wal/wal.go]: last_index? entry_index?

File: wal/wal.go [hunk 3]
```diff
@@ -49,8 +51,15 @@ var (
 	ErrSnapshotMismatch = errors.New("wal: snapshot mismatch")
 	ErrSnapshotNotFound = errors.New("wal: snapshot not found")
 	crcTable            = crc32.MakeTable(crc32.Castagnoli)
+
+	// metrics
+	syncHdr *metrics.Histogram
 )
 
+func init() {
+	syncHdr = metrics.NewHistogram("wal.syncInMicrosecond", 0, 10*int64(time.Second/time.Microsecond), 3)
```
Comment (yichengq) [inline wal/wal.go]: can this do in var part?

Comment (xiang90) [inline wal/wal.go]: i want to keep everything expvar initizaltion related in init

File: wal/wal.go (hunk 1 above)
Comment (xiang90) [inline wal/wal.go]: sync is an op. it is self-explained.

Comment (yichengq) [inline wal/wal.go]: k

**Comment by xiang90** [comment]
@yichengq update

File: wal/wal.go [hunk 4]
```diff
@@ -362,12 +376,14 @@ func (w *WAL) Close() error {
...
 
 func (w *WAL) saveEntry(e *raftpb.Entry) error {
+	metrics.Counter("wal.saveEntryOps").Add()
 	b := pbutil.MustMarshal(e)
 	rec := &walpb.Record{Type: entryType, Data: b}
 	if err := w.encoder.encode(rec); err != nil {
 		return err
 	}
 	w.enti = e.Index
+	metrics.Gauge("wal.lastIndexSaved").Set(int64(w.enti))
```
Comment (yichengq) [inline wal/wal.go]: Could we write it to
```
metrics.Gauge("wal.lastIndexSaved").SetFunc(func() int64 { return (int64)w.enti })
```
?

**Comment by yichengq** [comment]
lgtm

**Pull request merged**